epaxos/src/execute.rs

```diff
@@ -0,0 +1,169 @@
+use std::{
+    cmp::Ordering::{Equal, Greater, Less},
+    collections::{HashMap, VecDeque},
+};
+
+use petgraph::{
+    algo::tarjan_scc,
+    graph::{DiGraph, NodeIndex},
+};
+use tokio::sync::mpsc;
+
+use crate::types::{Command, CommandExecutor, InstanceSpace, InstanceStatus, SharedInstance};
+
+pub(crate) struct Executor<C, E>
+where
+    C: Command + Clone + Send + Sync + 'static,
+    E: CommandExecutor<C>,
+{
+    space: InstanceSpace<C>,
+    cmd_exe: E,
+}
+
+impl<C, E> Executor<C, E>
+where
+    C: Command + Clone + Send + Sync + 'static,
+    E: CommandExecutor<C> + Clone + Send + Sync + 'static,
+{
+    pub(crate) fn new(space: InstanceSpace<C>, cmd_exe: E) -> Self {
+        Self { space, cmd_exe }
+    }
+    // TODO:
+    // Wait missing instance and invalid state instance
+    pub(crate) async fn execute(&self, mut recv: mpsc::Receiver<SharedInstance<C>>) {
+        // A inifite loop to pull instance to execute
+        loop {
+            let ins = recv.recv().await;
+
+            if ins.is_none() {
+                // The channel has been closed, stop recving.
+                break;
+            }
+
+            let space_clone = Clone::clone(&self.space);
+            let executor_clone = Clone::clone(&self.cmd_exe);
+            tokio::spawn(async move {
+                let (scc, g) = build_scc(space_clone, ins.unwrap()).await;
+                execute(scc, g, executor_clone).await;
+            });
+        }
+    }
+}
+
+async fn execute<C, E>(scc: Vec<Vec<NodeIndex>>, g: DiGraph<SharedInstance<C>, ()>, cmd_exe: E)
+where
+    C: Command + Clone + Send + Sync + 'static,
+    E: CommandExecutor<C> + Clone + Send + Sync,
+{
+    for each_scc in scc {
+        let ins_vec = each_scc.iter().map(|index| &g[*index]);
+
+        let mut sort_vec = Vec::with_capacity(each_scc.len());
+        for (id, ins) in ins_vec.enumerate() {
+            let ins_read = ins.get_instance_read().await;
+            let ins_read = ins_read.as_ref().unwrap();
+            sort_vec.push((id, (ins_read.id.replica, ins_read.seq)));
+        }
+
+        sort_vec.sort_by(|a, b| {
+            match a.1 .1.partial_cmp(&b.1 .1) {
```

Comment:
Is the extra space between `a.1` and `1` intended?

Reply:
this space should be deleted

Reply:
it's the result of the formatter. Let's keep it.